naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java

```diff
@@ -811,7 +811,15 @@ public Instance getInstance(String namespaceId, String serviceName, String clust
             if (UtilsAndCommons.UPDATE_INSTANCE_ACTION_REMOVE.equals(action)) {
                 instanceMap.remove(instance.getDatumKey());
             } else {
-                instance.setInstanceId(instance.generateInstanceId(currentInstanceIds));
+                String instanceIdGenerator = instance.getInstanceIdGenerator();
+                if (Constants.SNOWFLAKE_INSTANCE_ID_GENERATOR.equalsIgnoreCase(instanceIdGenerator)) {
+                    Instance oldInstance = instanceMap.get(instance.getDatumKey());
+                    if (oldInstance != null && oldInstance.isSnowflakeId()) {
+                        instance.setInstanceId(oldInstance.getInstanceId());
+                    } else {
+                        instance.setInstanceId(generateSnowflakeInstanceId(currentInstanceIds));
+                    }
+                }
```

Comment:
Bad implementation I think. If oldInstance exist, only use instance id of old instance, can't change instanceId.

The original implementation also has some problem

Reply:
some situation.
1. old instance didn't use snowflake generator, new instance use snowflake generator. New instance id should be snowflake Id.
2.old instance use snowflake generator, new instance isn't use  snowflake generator. New instance id shouldn't remains old instance Id. 
3.old instance and new instance all use snowflake generator, new instance should remains old instance Id.

Reply:
1. Must deregister instance and register again.

2. Same as the first

3. yes it is.

Reply:
Instance id is a binder like ip and port (so default id generate from ip and port), If user want to change, they **must** deregister the older and register new. Can't update it directly.

Reply:
> Instance id is a binder like ip and port (so default id generate from ip and port), If user want to change, they **must** deregister the older and register new. Can't update it directly.

agree